pandas/tests/indexes/test_multi.py

```diff
@@ -2515,7 +2515,34 @@ def test_reconstruct_remove_unused(self):
         # idempotent
         result2 = result.remove_unused_levels()
         tm.assert_index_equal(result2, expected)
-        assert result2 is result
+        assert result2.is_(result)
+
+    def test_remove_unused_levels_large(self):
+        # GH16556
+
+        def check(first_type=None, second_type=None):
+            size = 1 << 16
+            first = np.random.randint(0, 1 << 13, size)
+            if first_type is not None:
+                first = first.astype(first_type)
+            second = np.random.randint(0, 1 << 10, size)
+            if second_type is not None:
+                second = second.astype(second_type)
+            third = np.random.rand(size)
+            df = DataFrame(dict(first=first, second=second, third=third))
+            df = df.groupby(['first', 'second']).sum()
+            df = df[df.third < 0.1]
+
+            result = df.index.remove_unused_levels()
+            assert len(result.levels[0]) < len(df.index.levels[0])
+            assert len(result.levels[1]) < len(df.index.levels[1])
+            assert result.equals(df.index)
+
+            expected = df.reset_index().set_index(['first', 'second']).index
+            assert result.equals(expected)
```

Comment:
I think I should replace this with the stronger `tm.assert_index_equal()`. I plan to do so when updating this test based on what you tell me to do about the non-determinism.